modules/runtime/viewer/pygame_viewer.py

```diff
@@ -32,36 +39,55 @@ def __init__(self, params=None, **kwargs):
 
+    def drawMap(self, map):
+        if self.map_surface == None:
+            # find boundary to create fixed size pygame surface
+            self.map_min_boundary = np.full(2, np.inf)
+            self.map_max_boundary = np.full(2, -np.inf)
+            lanes_np = []
+            lanes_dashed = []
+
+            for _, road in map.get_roads().items():
+                for lane_section in road.lane_sections:
+                    for _, lane in lane_section.get_lanes().items():
+                        lane_np = lane.line.toArray()
+                        lanes_np.append(lane_np)
+                        lanes_dashed.append(lane.road_mark.type == RoadMarkType.broken
+                                            or lane.road_mark.type == RoadMarkType.none)
+
+                        self.map_min_boundary = np.minimum(self.map_min_boundary, np.amin(lane_np, axis=0))
+                        self.map_max_boundary = np.maximum(self.map_max_boundary, np.amax(lane_np, axis=0))
+
+            self.map_size = self.map_max_boundary-self.map_min_boundary
+            # the size needed to be scaling larger for better visualization
+            # as a pygame surface is stored as pixels
+            self.map_surface = pg.Surface(tuple(self.map_size*self.screen_map_ratio))
+            self.map_surface.fill((255, 255, 255))
+
+            for lane_np, lane_dashed in zip(lanes_np, lanes_dashed):
+                pg.draw.aalines(self.map_surface, self.getColor(self.color_lane_boundaries), False, self.mapToCamera(lane_np), 3)
+
+        camera_coordinate = self.mapToCamera(np.array([self.dynamic_world_x_range[0], self.dynamic_world_y_range[1]]))
+        camera_view_range = np.array([np.diff(self.dynamic_world_x_range)[0], np.diff(self.dynamic_world_y_range)[0]])*self.screen_map_ratio
+        self.screen.blit(self.map_surface, (0, 0), tuple(np.around(np.concatenate((camera_coordinate, camera_view_range)))))
+
     def drawPoint2d(self, point2d, color, alpha):
         if self.screen is None:
             return
-        pg.draw.circle(self.screen, self.getColor(color),
-                       self.pointToPG(point2d), 1, 0)
+        pg.draw.circle(self.screen, self.getColor(color), self.pointsToCamera(point2d), 1, 0)
 
     def drawLine2d(self, line2d, color='blue', alpha=1.0, dashed=False):
         # TODO: enable dashed line
         if self.screen is None:
             return
-        line2d_np = line2d.toArray()
-        line_transformed_np = np.apply_along_axis(self.pointToPG, axis=1, arr=line2d_np)
-        for line_slice in np.ma.clump_unmasked(np.ma.masked_invalid(line_transformed_np)):
-            line_np = line_transformed_np[line_slice]
-            if len(line_np.tolist()) < 2:
-                continue
-            pg.draw.lines(
-                self.screen, self.getColor(color), False,
-                line_np.tolist(), 3)
+        line2d = self.pointsToCamera(line2d)
+        pg.draw.lines(self.screen, self.getColor(color), False, line2d, 3)
```

Comment:
can you elaborate on this chaange? why don't you need the transformation?

Reply:
Another main performance drop came from this implementation of transformation `np.apply_along_axis(self.pointToPG, axis=1, arr=...)` and the function calls following by this.
So I decide to using native numpy elementwise calculations to do the transformation with these two function
https://github.com/bark-simulator/bark/blob/7bc54de159edf12368e0994be1a481d743fccfa5/modules/runtime/viewer/pygame_viewer.py#L152-L162